fx/nnc_compile.py

```diff
@@ -0,0 +1,448 @@
+# This example is provided only for explanatory and educational purposes. The
+# underlying APIs may change and this tutorial may break.
+
+# Compiling FX models to NNC (Neural Network Compiler)
+######################################################
+# The goal of this file is to demonstrate an end to end example of using FX to
+# lower a PyTorch model to a backend codegen compiler. In this example, we will
+# be using NNC
+# (https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/tutorial.cpp).
+# If you're unfamiliar with NNC, the general design is strongly inspired by TVM
+# and Halide.
+#
+# To do so, this example contains two FX passes.
+# The first one is a decomposition pass that normalizes and decomposes PyTorch
+# operations (such as addmm). Using a pass like this allows us to reduce the
+# amount of lowerings we need to write. Instead of needing to specifically
+# write a lowering for addmm, we can decompose addmm and lower its constituent
+# operations.
+# The second one is the actual lowering pass itself. In this case, we will need
+# to convert each PyTorch operation we encounter into the corresponding NNC
+# `TensorExpr`.
+#
+# Writing these two passes, `decompose` and `nnc_compile`, are fairly similar.
+# In both cases, we re-interpret each operation in the FX graph to construct an
+# entirely new representation. In the decomposition pass, we either copy the
+# operation as-is into the new graph, or we use `Proxy` objects to decompose
+# the operation. This is an extension of the example presented here:
+# https://pytorch.org/docs/master/fx.html#proxy-retracing
+#
+# In the lowering pass, a similar principle applies. However, instead of using
+# `Proxy` objects to rewrite our op in other PyTorch ops, we do the translation
+# ourselves. In addition, since this is not a source-to-source transformation,
+# we return a somewhat hacky function that passes in the module attributes to
+# the NNC callable.
+#
+# Results
+######################################
+# Using NNC (which compiles directly to LLVM), we can compile a fairly small PyTorch model and compare performnance between NNC, PyTorch Eager, and Static Runtime.
+#
+# NNC time:  0.0066373348236083984
+# PyTorch time 0.025979042053222656
+# Static Runtime time 0.011004209518432617
```

Comment:
Maybe put your `lscpu` in a gist so people know what machine these results come from

Reply:
Also, is Static Runtime public at all? If not, remove references to it

Reply:
Bram has made references to it on the PyTorch dev discussion forum, so I assume it's public.

Reply:
I put my CPU version